# Practica 1. 

## Apartado 1: Particionado
Análisis de las dos estrategias de particionado propuestas: simple, y cruzada, para los conjuntos propuestos: german y tic-tac-toe. El análisis consiste en una descripción de los índices de train y test devueltos por cada uno de los métodos de particionado, junto con un comentario sobre las ventajas/desventajas de cada uno de ellos.

In [3]:
import numpy as np
from sklearn import preprocessing
from sklearn.naive_bayes import GaussianNB

from datos import *
from EstrategiaParticionado import *
#from Clasificador import *

#Este hay que cambiar a funciones de clasificador
from naive_bayes_functions import *


Simple validation assert: 0.3
Cross validation assert: [0.732, 0.748, 0.752, 0.748]


__Validación simple__: Este metodo de particionado divide los datos en 2 partes dada la proporción de datos test y train. Antes de la división los ids de las filas están mezcladas de manera aleatoria para asegurarse que los datos en train and test no esten en el mismo orden que  en el conjunto de datos original (por si están ordenados por clases o algún atributo.

__Ventajas__: Implementación simple y rápida

__Desventajas__: Si el conjunto de datos es pequeño quedan pocos datos para la validación. 

In [17]:
dataset=Datos('./ConjuntosDatos/tic-tac-toe.data') 

rows_number=dataset.datos.shape[0]

#Simple validation
test_proportion=0.3
line_ids=validacion_simple(rows_number,test_proportion)
line_ids_test=line_ids['Test']
line_ids_train=line_ids['Train']

print("Test: ", line_ids_test)

Test:  [0, 2, 3, 4, 5, 6, 10, 14, 16, 24, 26, 27, 29, 34, 37, 39, 48, 52, 55, 56, 60, 62, 66, 68, 71, 72, 78, 80, 81, 84, 95, 96, 97, 99, 100, 103, 104, 106, 107, 109, 117, 119, 120, 121, 128, 130, 131, 132, 135, 136, 137, 138, 141, 147, 149, 151, 154, 160, 166, 167, 171, 185, 192, 197, 204, 210, 215, 223, 228, 232, 236, 237, 238, 240, 241, 247, 251, 252, 253, 257, 258, 261, 263, 271, 274, 277, 279, 281, 284, 289, 292, 294, 296, 300, 306, 311, 312, 313, 321, 322, 323, 324, 327, 338, 339, 344, 356, 358, 360, 363, 364, 367, 368, 374, 378, 387, 390, 393, 395, 396, 399, 403, 410, 411, 413, 421, 422, 428, 432, 437, 439, 440, 442, 445, 454, 463, 465, 466, 469, 473, 480, 482, 488, 490, 491, 493, 504, 506, 507, 508, 509, 515, 516, 517, 521, 522, 526, 536, 538, 541, 548, 550, 552, 553, 557, 566, 569, 571, 574, 576, 581, 582, 589, 590, 591, 592, 593, 599, 616, 617, 618, 623, 626, 629, 633, 634, 636, 638, 641, 646, 651, 655, 656, 658, 661, 666, 667, 670, 675, 676, 684, 687, 690, 694, 696, 700, 70

Para extraer los datos se utiliza el metodo extraeDatos:

In [16]:
train=dataset.extraeDatos(line_ids_train)
test=dataset.extraeDatos(line_ids_test)
print(test)

[['x' 'x' 'x' ... 'o' 'x' 'positive']
 ['x' 'x' 'x' ... 'b' 'o' 'positive']
 ['x' 'x' 'x' ... 'o' 'b' 'positive']
 ...
 ['x' 'o' 'x' ... 'x' 'o' 'negative']
 ['o' 'x' 'x' ... 'x' 'x' 'negative']
 ['o' 'x' 'o' ... 'o' 'x' 'negative']]


__Validación cruzada__: En este tipo de validación el conjunto de datos de divide de manera aleatoria en N grupos, y uno de los grupos se utiliza para test y otros N-1 para train. De esta manera generamos N diferentes subconjuntos de train/test.
![alt text](K-fold_cross_validation.jpg "K-Fold Cross Validation")

__Ventajas__:Se utiliza el conjunto completo de datos para entrenamiento y para la validación y se puede generar más metricas para ver si el modelo el robusto.

__Desventajas__: El algoritmo hay que repetir N veces, lo que conlleva mayor coste computacional.

In [21]:
partitions=4
line_ids=validacion_cruzada(rows_number,partitions)
print("Test ids con segundo grupo asignado como test y train el resto:")
print(line_ids[1]['Test'])

Test ids con segundo grupo asignado como test y train el resto:
[1, 2, 3, 4, 13, 18, 31, 34, 39, 41, 48, 50, 57, 61, 65, 66, 69, 72, 74, 75, 89, 91, 93, 94, 98, 101, 105, 109, 112, 120, 127, 128, 129, 130, 138, 139, 158, 161, 169, 173, 175, 178, 182, 184, 185, 191, 193, 195, 199, 202, 204, 206, 210, 214, 220, 221, 223, 226, 236, 238, 239, 241, 243, 245, 246, 248, 251, 256, 259, 265, 278, 280, 286, 287, 291, 292, 295, 296, 301, 304, 310, 314, 326, 332, 335, 336, 338, 343, 350, 355, 356, 358, 360, 364, 365, 367, 368, 370, 380, 389, 390, 395, 397, 398, 408, 409, 412, 414, 416, 422, 428, 430, 435, 437, 441, 442, 447, 449, 450, 453, 455, 467, 470, 472, 473, 474, 476, 480, 482, 484, 489, 497, 504, 507, 512, 513, 514, 517, 523, 535, 536, 547, 552, 553, 556, 568, 571, 572, 581, 593, 594, 595, 602, 604, 606, 617, 621, 623, 630, 633, 636, 638, 639, 645, 646, 647, 652, 662, 666, 668, 675, 676, 683, 688, 694, 697, 698, 702, 707, 708, 713, 726, 729, 731, 737, 738, 742, 745, 748, 754, 756, 759, 761,

## Apartado 2: Naive-Bayes
Tabla con los resultados de la ejecución para los conjuntos de datos analizados (tic-tac-toe y german). Considerar los dos tipos de particionado.
Los resultados se refieren a las tasas de error/acierto y deben incluirse tanto con la corrección de Laplace como sin ella. Se debe incluir tanto el promedio de error para las diferentes particiones como su desviación típica. Es importante mostrar todos los resultados agrupados en una tabla para facilitar su evaluación.
Breve análisis de los resultados anteriores.

In [17]:
tictac=Datos('./ConjuntosDatos/tic-tac-toe.data') 

rows_tictac=tictac.datos.shape[0]

test_proportion=0.3

line_ids_tic=validacion_simple(rows_tictac,test_proportion)
line_ids_test_tic=line_ids_tic['Test']
line_ids_train_tic=line_ids_tic['Train']

train_tic=tictac.extraeDatos(line_ids_train_tic)
test_tic=tictac.extraeDatos(line_ids_test_tic)

print("Tic tac toe error, laplace =True", validacion(test_tic,tictac,train_tic, True))
print("Tic tac toe error, laplace =False", validacion(test_tic,tictac,train_tic, False))


german=Datos('./ConjuntosDatos/german.data') 

rows_german=german.datos.shape[0]

test_proportion=0.3
#Aqui tienen que estar las funciones de validacion de clasificador
line_ids_german=validacion_simple(rows_german,test_proportion)
line_ids_test_german=line_ids_german['Test']
line_ids_train_german=line_ids_german['Train']

train_german=german.extraeDatos(line_ids_train_german)
test_german=german.extraeDatos(line_ids_test_german)


print("German error, laplace=True", validacion(test_german,german,train_german, True))
print("German error, laplace=False", validacion(test_german,german,train_german, False))

Tic tac toe error, laplace =True 0.344
Tic tac toe error, laplace =False 0.344
German error, laplace=True 0.272
German error, laplace=False 0.272


In [22]:
##Cruzada
partitions=4
#Tic tac
line_ids=validacion_cruzada(rows_tictac,partitions)
assert_cross_tictac=[]
assert_cross_tictac_laplace=[]
for i in range(partitions):
    line_ids_test=line_ids[i]['Test']
    line_ids_train=line_ids[i]['Train']
    train=tictac.extraeDatos(line_ids_train)
    test=tictac.extraeDatos(line_ids_test)
    assert_cross_tictac.append(validacion(test,tictac,train, False))
    assert_cross_tictac_laplace.append(validacion(test,tictac,train, True))
print("Tic-tac-toe")
print("Without laplace", assert_cross_tictac, "Mean: ", np.mean(assert_cross_tictac), "Std:", np.std(assert_cross_tictac))
print("With Laplace", assert_cross_tictac_laplace, "Mean: ", np.mean(assert_cross_tictac_laplace), "Std:", np.std(assert_cross_tictac_laplace))



line_ids=validacion_cruzada(rows_german,partitions)
assert_cross_german=[]
assert_cross_german_laplace=[]
for i in range(partitions):
    line_ids_test=line_ids[i]['Test']
    line_ids_train=line_ids[i]['Train']
    train=german.extraeDatos(line_ids_train)
    test=german.extraeDatos(line_ids_test)
    assert_cross_german.append(validacion(test,german,train, False))
    assert_cross_german_laplace.append(validacion(test,german,train, True))
print("German")
print("Without laplace", assert_cross_german, "Mean: ", np.mean(assert_cross_german), "Std:", np.std(assert_cross_german))
print("With Laplace", assert_cross_german_laplace, "Mean: ", np.mean(assert_cross_german_laplace), "Std:", np.std(assert_cross_german_laplace))

Tic-tac-toe
Without laplace [0.352, 0.264, 0.296, 0.248] Mean:  0.29 Std: 0.03974921382870357
With Laplace [0.352, 0.264, 0.296, 0.248] Mean:  0.29 Std: 0.03974921382870357
German
Without laplace [0.252, 0.26, 0.28, 0.244] Mean:  0.259 Std: 0.013379088160259664
With Laplace [0.252, 0.26, 0.28, 0.244] Mean:  0.259 Std: 0.013379088160259664


__Error con validación simple__

|Dataset|tic-tac-toe|german|
|-------|-----------|------|
|Error (Laplace)| 0.344| 0.272|
|Error (No Laplace)| 0.344| 0.272|

__Error con validación cruzada__

|Dataset|tic-tac-toe|german|
|-------|-----------|------|
|Mean Error| 0.29 | 0.259|
|Std| 0.0397| 0.0134|


El error aplicando y no aplicando Laplace para estos conjuntos de datos es igual, no influye mucho a la precisión de clasificador. 
El error de validación cruzada es más bajo porque hemos utilzado más datos para entrenamiento en cada de sus iteraciones (75% vs 70% en validación simple).

## Apartado 3: Scikit-Learn
Incluir los mismos resultados que en el apartado 2 pero usando los métodos del paquete scikit-learn. Comparar y analizar los resultados.

## Apartado 4: Evaluación de hipótesis mediante Análisis ROC
Matriz de confusión y diagramas del clasificador en el espacio ROC